libsolidity/formal/BMC.cpp

```diff
@@ -149,11 +148,14 @@ void BMC::endVisit(ContractDefinition const& _contract)
 
 bool BMC::visit(FunctionDefinition const& _function)
 {
+	// Free functions need to be visited in the context of a contract.
+	if (!m_currentContract)
+		return false;
+
 	auto contract = dynamic_cast<ContractDefinition const*>(_function.scope());
```

Comment:
Maybe move just before first use (before `if (contract && find(hierarchy.begin(), hierarchy.end(), contract) == hierarchy.end())`)

Also, below there is a code like this:
```
if (_function.isConstructor())
		inlineConstructorHierarchy(dynamic_cast<ContractDefinition const&>(*_function.scope()));
```
Can we use `*contract` here instead of dynamic cast? Also, we should have assertion here as the one on the beginning of a function was removed

Reply:
Moved closer but still before `hierarchy` since that's also used in the if. Also added `solAssert` and replaced cast